### PR TITLE
Added support for loading Python at runtime (supports both Python 3 and 2)

### PR DESCRIPTION
**Pull request by pvieito**
Added new PythonLibrary struct to load Python library and symbols dynamically at runtime

- This adds support for both Python 3 and 2 with the same codebase.
- You can force a version of Python with the PYTHON_VERSION environment variable, a Python library path with PYTHON_LIBRARY and enable logging with PYTHON_LOADER_LOGGING.
- The implementation comes from the PythonKit Swift package (https://github.com/pvieito/PythonKit) removing the Foundation dependency.

### CODE REVIEW

File: stdlib/public/Python/PythonLibrary.swift
Comment (xwu) [inline stdlib/public/Python/PythonLibrary.swift]: Would this be more elegant if there were simply an array of tuples called `supportedVersions: [(major: Int, minor: Int?)]`?

Such a tuple could also be manually populated (i.e., `[(3, nil), (3, 8), (3, 7), /* ... */]`). There's no need for supporting nonexistent Python versions (indeed, that would be an oxymoron).

File: stdlib/public/Python/PythonLibrary.swift
Comment (rxwei) [inline stdlib/public/Python/PythonLibrary.swift]: Make the extension private and drop ‘private’ from members.

File: stdlib/public/Python/PythonLibrary.swift
Comment (rxwei) [inline stdlib/public/Python/PythonLibrary.swift]: Avoid using “get” or “set” in function names.

Plus, I think this can be a computed property.

File: stdlib/public/Python/PythonLibrary.swift
Comment (rxwei) [inline stdlib/public/Python/PythonLibrary.swift]: Is this used in other files? If not, make the extension private.

File: stdlib/public/Python/PythonLibrary.swift
Comment (rxwei) [inline stdlib/public/Python/PythonLibrary.swift]: Should this be private too?

File: stdlib/public/Python/PythonLibrary.swift
Comment (rxwei) [inline stdlib/public/Python/PythonLibrary.swift]: Argument label “message” doesn’t add any clarity. Why not just log(_:)?

File: stdlib/public/Python/PythonLibrary.swift
Comment (rxwei) [inline stdlib/public/Python/PythonLibrary.swift]: “signature” is an overloaded word. I’d suggest using “type:”.

File: stdlib/public/Python/PythonLibrary.swift
Comment (pvieito) [reply stdlib/public/Python/PythonLibrary.swift]: Removing support for "nonexistent" Python versions means that old Swift distributions (or even any distributed executable that statically linked the Python module) would not work with any 3.x Python version released after it, even though the Python ABI is marked as stable for all the Python 3 lifetime. Also, it would require maintaining the list and deciding when it's time to add new versions.

Take in account that while in Windows the Python distribution includes both `python3.dll` and `python37.dll` in macOS and Linux we have to specify the exact minor version to load it (`Python.framework/Versions/3.7/Python` or `libpython3.7m.so`).

Comment (xwu) [reply stdlib/public/Python/PythonLibrary.swift]: If the ABI is stable for all versions of Python 3, then is there a need for a list of supported minor versions?

Comment (pvieito) [reply stdlib/public/Python/PythonLibrary.swift]: @xwu Read the second paragraph.

**Review by rxwei — APPROVED** [review]
LGTM. Thanks!

We will also need an API that enables the user to select a Python version. Something like: `Python.useVersion(3, 6)`. It can be added in a separate patch.

File: stdlib/public/Python/PythonLibrary.swift
Comment (xwu) [reply stdlib/public/Python/PythonLibrary.swift]: I'm afraid I don't understand. Why do we need a list of supported minor versions?

Comment (xwu) [reply stdlib/public/Python/PythonLibrary.swift]: I understand that. What I don't understand is why we need a static array of "supported" minor versions if support is not contingent on the minor version.

**Comment by pvieito** [comment]
The function should be on `PythonLibrary` as merely using `Python` already calls `Py_Initialize()`.

Something like this would do the trick:

```swift
// Methods of `PythonLibrary` required to set the Python version.
extension PythonLibrary {
  private static func useVersion(_ version: PythonVersion) {
    PythonLibrary.Environment.version.set(version.versionString)
  }
  
  public static func useVersion(_ major: Int, _ minor: Int?) {
    let version = PythonVersion(major: major, minor: minor)
    self.useVersion(version)
  }
}
```

Example:

```swift
import Python

PythonLibrary.useVersion(2) // Should be called before any references to `Python`
let sys = try Python.attemptImport("sys")
print(sys.version)
```

**Comment by rxwei** [comment]
IMO, using a `PythonVersion` type is unnecessary complexity and is not intuitive. I don't think we need any extra types to represent the python version, or even separate major from minor from an API perspective. The dlopen-based API is fully dynamic, extra types won't give us more safety than just using a `Float`. How about just starting with a very simple API:

```swift
public extension PythonLibrary {
  static func useVersion(_ version: Float) {
    ... // do version tuple conversion here and set environment
  }
}
```

**Comment by pvieito** [comment]
Take in account that the `PythonVersion` struct is private and it is also used on `loadPythonLibrary(at:version:)`. Also, `Float` is a really bad idea: `3.10` vs `3.1`.

**Comment by rxwei** [comment]
> Also, `Float` is a really bad idea: `3.10` vs `3.1`.

Ok, I'm sold.

> The public method will be: `static func useVersion(_ major: Int, _ minor: Int? = nil)`.

Adding a single method for now and making `PythonVersion` private sounds good to me.

**Comment by pvieito** [comment]
> Adding a single method for now and making `PythonVersion` private sounds good to me.

Great, done! 👍🏻

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
CI is dead. Gotta wait a while.

**Comment by rxwei** [comment]
Could you add new files to `CMakeLists.txt`?

**Comment by pvieito** [comment]
@rxwei Should I also remove the `CPython` module?

**Comment by pvieito** [comment]
@swift-ci please test tensorflow

**Comment by dan-zheng** [comment]
@pvieito Yes, please remove `CPython`. I see you've already done so. :)

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
I believe the `unknown expression syntax exists in the source` errors are related to libSyntax verification, which is triggered by the flags `-Xfrontend -verify-syntax-tree`.

Could you try building PythonKit on Linux with those extra flags? Unfortunately, this seems like a libSyntax deficiency. [`TypeNodes.py`](https://github.com/apple/swift/blob/tensorflow/utils/gyb_syntax_support/TypeNodes.py) may not yet have support for function conventions.

@nkcsgexi: could you confirm whether libSyntax supports function conventions like `@convention(c)`?

**Comment by pvieito** [comment]
@dan-zheng Exactly:

```
PythonKit$ swift build -Xswiftc -Xfrontend -Xswiftc -verify-syntax-tree   
Compile Swift Module 'PythonKit' (4 sources)
Compile Swift Module 'LoggerKit' (12 sources)
/Developer/PythonKit/PythonKit/PythonLibrary+Symbols.swift:39:10: error: unknown expression syntax exists in the source
  type: (@convention(c) () -> ()).self
         ^
/Developer/PythonKit/PythonKit/PythonLibrary+Symbols.swift:44:10: error: unknown expression syntax exists in the source
  type: (@convention(c) (PyObjectPointer?) -> ()).self
         ^
/Developer/PythonKit/PythonKit/PythonLibrary+Symbols.swift:49:10: error: unknown expression syntax exists in the source
  type: (@convention(c) (PyObjectPointer?) -> ()).self
         ^
...
```

**Comment by dan-zheng** [comment]
@pvieito Hmm, that's unfortunate. Unless we can disable `-verify-syntax-tree` for the Python module, this PR is blocked by libSyntax support for this usage of `@convention`.

@nkcsgexi: Could you please provide some advice on how to add libSyntax support for this usage of `@convention`? (Basically, how to resolve [SR-9316](https://bugs.swift.org/browse/SR-9316)). That would be greatly appreciated!

**Comment by rxwei** [comment]
The libSyntax issue is a bit surprising. There's no instance of getting the metatype of a `@convention(c)` function in stdlib. But let's not get blocked on this. You can work around this by first making `type` a default argument:

```swift
static func loadSymbol<T>(
  name: String, legacyName: String? = nil, type: T.Type = T.self
) -> T
```

Then change your call sites to provide a contextual type.

```swift
let PyList_New: @convention(c) (Int) -> PyObjectPointer? =
  PythonLibrary.loadSymbol(name: "PyList_New")
```

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: stdlib/public/Python/PythonLibrary.swift
Comment (xwu) [reply stdlib/public/Python/PythonLibrary.swift]: We can work on this in a follow-up.

There will never be a version of Python 2 after 2.7 for instance, and it makes no sense to look for one; if "Python 2.30" existed, we certainly wouldn't want to support it. Likewise, is Python 2.0 really supported? This is all to say that iterating over the same minor versions for every major version is not ideal.

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
It's still failing possibly because of incorrectly configured cmake. You mentioned you couldn't test it on your machine, is that right? If so, I can take it from here.

**Comment by rxwei** [comment]
@pvieito I'm currently debugging some failing tests:

```console
[ RUN      ] PythonRuntime.CheckVersion
stdout>>> check failed at /Users/rxwei/Development/Swift/swift-tf-source/swift/test/Python/python_runtime.swift, line 14
stdout>>> expected: "2.7." (of type Swift.String)
stdout>>> actual: "3.7." (of type Swift.Substring)
[     FAIL ] PythonRuntime.CheckVersion
[ RUN      ] PythonRuntime.PythonList
stdout>>> check failed at /Users/rxwei/Development/Swift/swift-tf-source/swift/test/Python/python_runtime.swift, line 22
stdout>>> expected: <type 'list'> (of type Python.PythonObject)
stdout>>> actual: <class 'list'> (of type Python.PythonObject)
[     FAIL ] PythonRuntime.PythonList
[ RUN      ] PythonRuntime.PythonDict
[       OK ] PythonRuntime.PythonDict
[ RUN      ] PythonRuntime.Range
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.Range
[ RUN      ] PythonRuntime.PartialRangeFrom
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.PartialRangeFrom
[ RUN      ] PythonRuntime.PartialRangeUpTo
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.PartialRangeUpTo
[ RUN      ] PythonRuntime.BinaryOps
[       OK ] PythonRuntime.BinaryOps
[ RUN      ] PythonRuntime.Comparable
stderr>>> Fatal error: 'try!' expression unexpectedly raised an error: Python exception: '<' not supported between instances of 'int' and 'str': file /Users/rxwei/Development/Swift/swift-tf-source/swift/stdlib/public/Python/Python.swift, line 1173
stderr>>> CRASHED: SIGILL
the test crashed unexpectedly
[     FAIL ] PythonRuntime.Comparable
[ RUN      ] PythonRuntime.Hashable
[       OK ] PythonRuntime.Hashable
[ RUN      ] PythonRuntime.RangeIteration
[       OK ] PythonRuntime.RangeIteration
[ RUN      ] PythonRuntime.Errors
[       OK ] PythonRuntime.Errors
[ RUN      ] PythonRuntime.Tuple
[       OK ] PythonRuntime.Tuple
[ RUN      ] PythonRuntime.MethodCalling
[       OK ] PythonRuntime.MethodCalling
[ RUN      ] PythonRuntime.ConvertibleFromPython
[       OK ] PythonRuntime.ConvertibleFromPython
[ RUN      ] PythonRuntime.PythonConvertible
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] PythonRuntime.PythonConvertible
PythonRuntime: Some tests failed, aborting
```

It seems to be loading Python 3 by default. However, even if I explicitly call `useVersion(2, 7)`, the `CheckVersion` test still produces 3.7.

```swift
PythonLibrary.useVersion(2, 7)

PythonRuntimeTestSuite.test("CheckVersion") {
  let sysModule = Python.import("sys")
  let version = String(sysModule.version)!
  expectEqual("2.7.", version.prefix(4))
}
```
```console
stdout>>> check failed at /Users/rxwei/Development/Swift/swift-tf-source/swift/test/Python/python_runtime.swift, line 14
stdout>>> expected: "2.7." (of type Swift.String)
stdout>>> actual: "3.7." (of type Swift.Substring)
```

**Comment by pvieito** [comment]
It seems that we have to reference the symbol:

```swift
guard isType(pythonObject, type: PySlice_Type) else { return nil }
```

Not its address to it:

```swift 
guard isType(pythonObject, type: &PySlice_Type) else { return nil } // Crashes
```

**Comment by rxwei** [comment]
> I suppose we should modify the symbol definition (to match the real Python C header) instead the calling site?

Symbol definitions are correct. It's just that we don't need these type variables to be `var`s anymore since they are addresses to the actual objects.

**Comment by pvieito** [comment]
Ok, great!

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by pvieito** [comment]
@rxwei Please, remove also the symbol `_Py_TrueStructb` (seems to be a copy-paste typo).

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux GPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux GPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux GPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by rxwei** [comment]
@swift-ci please test tensorflow macOS

**Comment by H-World** [comment]
@pvieito  我想知道在Mac上怎么使用指定版本的python，现在我使用PythonLibrary.useVersion(3)，会崩溃Python library not found. Set the PYTHON_LIBRARY environment variable with the path to a Python library.但是我又找不到方法去指定PYTHON_LIBRARY

**Pull request merged**
